packages/protocol/engine/congestioncontrol/icca/scheduler/scheduler.go

```diff
@@ -0,0 +1,610 @@
+		return err
+	}
+
+	s.ready(block)
+
+	return nil
+}
+
+// isEligible returns true if the given blockID has either been scheduled or confirmed.
+func (s *Scheduler) isEligible(block *Block) (eligible bool) {
+	return block.Scheduled() || s.isBlockAcceptedFunc(block.ID())
+}
+
+// isReady returns true if the given blockID's parents are eligible.
+func (s *Scheduler) isReady(block *Block) (ready bool) {
+	ready = true
+	block.ForEachParent(func(parent models.Parent) {
+		if parentBlock, parentExists := s.block(parent.ID); !parentExists || !s.isEligible(parentBlock) {
+			ready = false
+			return
+		}
+	})
+
+	return
+}
+
+// tryReady tries to set the given block as ready.
+func (s *Scheduler) tryReady(block *Block) {
+	if s.isReady(block) {
+		s.Ready(block)
+	}
+}
+
+// updateChildren iterates over the direct children of the given blockID and
+// tries to mark them as ready.
+func (s *Scheduler) updateChildren(block *Block) {
+	for _, childBlock := range block.Children() {
+		if childBlockScheduler, childBlockExists := s.block(childBlock.ID()); childBlockExists {
+			s.tryReady(childBlockScheduler)
+		}
+	}
+}
+
+func (s *Scheduler) submit(block *Block) error {
+	if s.stopped.IsSet() {
+		return ErrNotRunning
+	}
+
+	// TODO: when removing the zero mana issuer solution, check if issuers have MinMana here
+	droppedBlocks, err := s.buffer.Submit(block, s.getAccessMana)
+	if err != nil {
+		return errors.Wrapf(err, "failed to submit %s", block.ID())
+	}
+
+	s.markAsDropped(droppedBlocks)
+
+	return nil
+}
+
+func (s *Scheduler) markAsDropped(droppedBlocks []*Block) {
+	for _, droppedBlock := range droppedBlocks {
+		s.Tangle.SetOrphaned(droppedBlock.Block.Block.Block, true)
+		s.Events.BlockDropped.Trigger(droppedBlock)
+	}
+}
+
+func (s *Scheduler) unsubmit(block *Block) {
+	s.buffer.Unsubmit(block)
+}
+
+func (s *Scheduler) ready(block *Block) {
+	s.buffer.Ready(block)
+}
+
+// block retrieves the Block with given id from the mem-storage.
+func (s *Scheduler) block(id models.BlockID) (block *Block, exists bool) {
+	if s.EvictionManager.IsRootBlock(id) {
+		tangleBlock, _ := s.Tangle.VirtualVoting.Block(id)
+
+		return NewBlock(tangleBlock, WithScheduled(true)), true
+	}
+
+	storage := s.blocks.Get(id.Index(), false)
+	if storage == nil {
+		return nil, false
+	}
+
+	return storage.Get(id)
+}
+
+// mainLoop periodically triggers the scheduling of ready blocks.
+func (s *Scheduler) mainLoop() {
+	defer s.ticker.Stop()
+
+loop:
+	for {
+		select {
+		// every rate time units
+		case <-s.ticker.C:
+			// TODO: pause the ticker, if there are no ready blocks
+			if block := s.schedule(); block != nil {
+				if block.SetScheduled() {
+					s.Events.BlockScheduled.Trigger(block)
+				}
+			}
+
+		// on close, exit the loop
+		case <-s.shutdownSignal:
+			break loop
+		}
+	}
+}
+
+func (s *Scheduler) schedule() *Block {
+	s.bufferMutex.Lock()
+	defer s.bufferMutex.Unlock()
+	s.EvictionManager.RLock()
+	defer s.EvictionManager.RUnlock()
```

Comment:
these two locks seem to be highly correlated. it's a bit of a mess to follow the code and the locking tbh

Reply:
That's true, but I don't think we only can have one. Eviction lock is used whenever we execute a publicly facing method that accesses block storage or `schedule` method so that eviction is not performed in the middle. 
Buffer mutex is used by any method that accesses or modifies the block, so it's true that they often go together, but not always.
Also eviction is mostly read locked while buffer is often write locked.